src/AccountServices.js

```diff
@@ -0,0 +1,257 @@
+    }).catch(err => {
+      intent.sendMessage(event.room_id, {
+        "body": "We couldn't verify this PIN :(. Maybe you typed it wrong or you"
+             + "might need to request it again.",
+        "msgtype": "m.text"
+      });
+      log.error("Handler.AccountServices", "OAuth Access Token Failed:%s", err);
+    });
+  });
+}
+
+AccountServices.prototype._oauth_getAccessToken = function (pin, client_data, id) {
+  return new Promise((resolve, reject) => {
+    if (!client_data || client_data.oauth_token == "" || client_data.oauth_secret == "") {
+      reject("User has no associated token request data");
+      return;
+    }
+    this._oauth.getOAuthAccessToken(
+      client_data.oauth_token,
+      client_data.oauth_secret,
+      pin,
+      (error, access_token, access_token_secret) =>{
+        if(error) {
+          reject(error.statusCode + ": " + error.data);
+          return;
+        }
+        client_data.access_token = access_token;
+        client_data.access_token_secret = access_token_secret;
+        var client = new Twitter({
+          consumer_key: this._app_auth.consumer_key,
+          consumer_secret: this._app_auth.consumer_secret,
+          access_token_key: client_data.access_token,
+          access_token_secret: client_data.access_token_secret
+        });
+        client.get("account/verify_credentials", (error, profile) =>{
+          if(error) {
+            log.error("Handler.AccountServices", `We couldn't authenticate with `
+          +`the supplied access token for ${id}. ${error}`);
+            reject("Twitter account could not be authenticated.");
+            return;
+          }
+          this._storage.set_twitter_account(id, profile.id, client_data).then(() =>{
+            resolve(profile);
+          }).catch(() => {
+            reject("Failed to store account information.")
+          })
+        });
+      });
+  });
+};
+
+/**
+ * _oauth_getUrl - Start the process of connecting an account to Twitter.
+ *
+ * @param  {type} id    The matrix user id wishing to authenticate.
+ * @return {Promise<string>}  A promise that will return an auth url or reject with nothing.
+ */
+AccountServices.prototype._oauth_getUrl = function (id) {
+  return new Promise((resolve, reject) => {
+    this._oauth.getOAuthRequestToken(
+       /* 'x_auth_access_type' is used to specify the access level.
+        * Valid options are:
+        * read - Read from the API
+        * write - read + make changes
+        * dm - read+write + be able to send/read direct messages
+        * */
+      {"x_auth_access_type": "dm"},
+      (error, oAuthToken, oAuthTokenSecret) => {
+        if(error) {
+          reject(error);
+        }
+        var data = {
+          oauth_token: oAuthToken,
+          oauth_secret: oAuthTokenSecret,
+          access_token: "",
+          access_token_secret: ""
+        };
+        this._storage.set_twitter_account(id, "", data).then(()=>{
```

Comment:
Yes, it's sort of a midway storage to let the bridge know that we have begun the linking process and are waiting for the pin to be returned.